### ISSUE TITLE
Hide completed progress bars during parallel downloads

### ISSUE THREAD

**Issue by dholth** [conda/conda#12210]
> This is amazing. However, it doesn't seem to always show the unfinished downloads and now I cannot see the "more hidden" progress (for quite a while) 😂

![image](https://user-images.githubusercontent.com/7588612/209442638-4393d75a-e47f-409a-9fe6-e1ae19014fa3.png)

_Originally posted by @yzhang-gh in https://github.com/conda/conda/issues/11608#issuecomment-1364545763_

**Comment by dholth** [comment]
It's tricky to control progress bars from multiple threads. We try to deal with them using the leave, position (appears to be unused/automatically handled), and `.close()` all at the end.

What happens if we close individual bars as each download completes? (Or one second after each download completes, so it doesn't immediately disappear?)

Might we need to coordinate changing the position of the other bars, when a bar is removed?

https://github.com/conda/conda/blob/main/conda/core/package_cache_data.py#L742

https://github.com/conda/conda/blob/main/conda/core/package_cache_data.py#L796

https://github.com/conda/conda/blob/main/conda/core/package_cache_data.py#L819

**Comment by jaimergp** [comment]
We might be at a point where `rich` can help simplify things. Some examples and related discussions:

- https://github.com/Textualize/rich/blob/master/examples/dynamic_progress.py
- https://github.com/Textualize/rich/blob/master/examples/downloader.py
- https://github.com/Textualize/rich/discussions/1500
- https://github.com/pypa/pip/issues/8698

**Comment by dholth** [comment]
Related issue to be created: tqdm appears to be still be active when the --json progress bar substitute is requested.

**Comment by dholth** [comment]
Could we adopt something like fsspec's callback interface https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.callbacks.Callback

**Comment by MrMarvel** [comment]
I want it so badly.
I wait over 10 minutes and doesn't know how much I need to wait because it doesn't show progress
![image](https://github.com/conda/conda/assets/24212110/e58b968b-d1a5-4de3-b12f-37f0f06a3146)

**Comment by dholth** [comment]
We would still love to fix this too, but the next release will include two quick improvements. https://github.com/conda/conda/issues/13248 which downloads the biggest package first (to speed up downloads, and so the one you're waiting for is on the top) and https://github.com/conda/conda/pull/13231 (so that conda can resume interrupted downloads).

**Comment by ansanyuan** [comment]
me too

**Comment by owlshrimp** [comment]
This would be an *enormous* quality-of-life improvement.

**Comment by jaimergp** [comment]
FWIW, there's a prototype plugin to replace the progress bars: https://github.com/conda-incubator/conda-rich. It's not conda installable yet, but you can hack your way in if you know what you are doing. This is NOT supported, so only do it in a non-production environment:

```
conda activate base
conda install rich
SETUPTOOLS_SCM_PRETEND_VERSION=25.11 pip install https://github.com/conda-incubator/conda-rich/archive/main.tar.gz --no-deps
conda config --set console rich
```

**Comment by owlshrimp** [comment]
just an idea, but would it make sense to simply give each thread less control over printing the progress bars, and just have them asynchronously notify a rendering thread of their progress? after all, it doesn't really matter if the display lags behind.

**Comment by dholth** [comment]
Yes, it would make sense. At the time we had trouble getting that to work with tqdm, but anyone is welcome to try their hand at it.

**Comment by dholth** [comment]
Not stale.
